#! /bin/sh /usr/share/dpatch/dpatch-run
## dhclient-pretty_print-off-by-one.dpatch by  <apollock@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: dhclient is incorrectly decoding the domain-search option, and only 
## DP: returning the first domain in the list
## DP: this patch from 4.0.0

@DPATCH@
diff -urNad dhcp3-3.1.1~/common/options.c dhcp3-3.1.1/common/options.c
--- dhcp3-3.1.1~/common/options.c	2008-05-14 22:36:12.000000000 -0700
+++ dhcp3-3.1.1/common/options.c	2008-06-14 13:09:22.845142436 -0700
@@ -1567,17 +1567,31 @@
 						pretty_domain(&op, endbuf-1,
 							      &nbp, nend);
 					} else {
+						/* ns_name_ntop() includes
+						 * a trailing NUL in its
+						 * count.
+						 */
 						count = MRns_name_ntop(
 								nbuff, op, 
 								(endbuf-op)-1);
 
-						if (count == -1) {
+						if (count <= 0) {
 							log_error("Invalid "
 								"domain name.");
 							break;
 						}
 
-						op += count;
+						/* Consume all but the trailing
+						 * NUL.
+						 */
+						op += count - 1;
+
+						/* Replace the trailing NUL
+						 * with the implicit root
+						 * (in the unlikely event the
+						 * domain name /is/ the root).
+						 */
+						*op++ = '.';
 					}
 				}
 				*op = '\0';
